src/Orleans/ApplicationParts/ApplicationPartManagerExtensions.cs

```diff
@@ -0,0 +1,147 @@
+            }
+
+            return result;
+        }
+
+        public static void AddApplicationPart(this ApplicationPartManager applicationPartManager, Assembly assembly)
+        {
+            if (applicationPartManager == null)
+            {
+                throw new ArgumentNullException(nameof(applicationPartManager));
+            }
+
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            applicationPartManager.AddApplicationPart(new AssemblyPart(assembly));
+        }
+
+        public static void AddApplicationPartsFromAppDomain(this ApplicationPartManager applicationPartManager, bool loadReferencedAssemblies = true)
+        {
+            if (applicationPartManager == null)
+            {
+                throw new ArgumentNullException(nameof(applicationPartManager));
+            }
+
+            var processedAssemblies = new HashSet<Assembly>();
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
+            {
+                if (processedAssemblies.Add(assembly) && loadReferencedAssemblies)
+                {
+                    AddReferencedAssemblies(assembly, processedAssemblies);
+                }
+
+                applicationPartManager.AddApplicationPart(assembly);
+            }
+
+            void AddReferencedAssemblies(Assembly asm, HashSet < Assembly > loadedAssemblies)
+            {
+                if (asm == null)
+                {
+                    throw new ArgumentNullException(nameof(asm));
+                }
+
+                if (loadedAssemblies == null)
+                {
+                    throw new ArgumentNullException(nameof(loadedAssemblies));
+                }
+
+                var referenced = asm.GetReferencedAssemblies();
+                foreach (var asmName in referenced)
+                {
+                    try
+                    {
+                        var refAsm = Assembly.Load(asmName);
+                        if (loadedAssemblies.Add(refAsm)) AddReferencedAssemblies(refAsm, loadedAssemblies);
+                    }
+                    catch
+                    {
+                        // Ignore loading exceptions.
+                    }
+                }
+            }
+        }
+
+        private static readonly string CoreAssemblyName = typeof(RuntimeVersion).Assembly.GetName().Name;
+        private static readonly string AbstractionsAssemblyName = typeof(IGrain).Assembly.GetName().Name;
+        private static readonly IEnumerable<string> NoReferenceComplaint = new[] {$"Assembly does not reference {CoreAssemblyName} or {AbstractionsAssemblyName}"};
+
+        public static void AddApplicationPartsFromDefaultDirectories(this ApplicationPartManager applicationPartManager)
+        {
+            var exeRoot = Path.GetDirectoryName(typeof(ApplicationPartManagerExtensions).GetTypeInfo().Assembly.Location);
```

Comment:
Is this correct? exeRoot will not be a folder, but a DLL path, so combining it with `Applications` will not be very nice.
Also, does this work in .NET Core? We've seen weird issues in our preview3 assembly loading that now requires to `publish` the build first, otherwise the assemblies run from weird locations. If we are taking this approach forward, then we should probably fix it (in case it wasn't, I didn't validate if it's the same approach as before).

Reply:
I believe it's correct: we're calling `Path.GetDirectoryName` on the result. This is used during tests and it seems to be working there.
As far as I know, publishing a build is the only valid way to deploy it. I'm not sure the best approach.

Should I perhaps add an extra parameter to specify paths?